src/software/ai/passing/pass_generator.cpp

```diff
@@ -1,381 +1,116 @@
-              [this](Pass p1, Pass p2) { return comparePassQuality(p1, p2); });
-
-    // Merge Passes That Are Similar
-    // We start by assuming that the most similar passes will be right beside each other,
-    // then iterate over the entire list, building a new list as we go by only adding
-    // elements when they are dissimilar enough from the last element we added
-    // NOTE: This flips the passes so they are sorted by increasing quality
-    passes_to_optimize = std::accumulate(
-        passes_to_optimize.begin(), passes_to_optimize.end(), std::vector<Pass>(),
-        [this](std::vector<Pass>& passes, Pass curr_pass) {
-            // Check if we have no passes, or if this pass is too similar to the
-            // last pass we added to the list
-            if (passes.empty() || !passesEqual(curr_pass, passes.back()))
-            {
-                passes.emplace_back(curr_pass);
-            }
-            return passes;
-        });
-
-    // Replace the least promising passes
-    passes_to_optimize.erase(
-        passes_to_optimize.begin() +
-            passing_config->getNumPassesToKeepAfterPruning()->value(),
-        passes_to_optimize.end());
-
-    // Generate new passes to replace the ones we just removed
-    int num_new_passes = passing_config->getNumPassesToOptimize()->value() -
-                         static_cast<int>(passes_to_optimize.size());
-    if (num_new_passes > 0)
-    {
-        std::vector<Pass> new_passes = generatePasses(num_new_passes);
-        // Append our newly generated passes to replace the passes we just removed
-        passes_to_optimize.insert(passes_to_optimize.end(), new_passes.begin(),
-                                  new_passes.end());
-    }
-}
+    ZonePassMap<ZoneEnum> optimized_passes;
 
-void PassGenerator::saveBestPass()
-{
-    // Take ownership of the best_known_pass for the duration of this function
-    std::lock_guard<std::mutex> best_known_pass_lock(best_known_pass_mutex);
-
-    // Sort the passes by decreasing quality
-    std::sort(
-        passes_to_optimize.begin(), passes_to_optimize.end(),
-        [this](auto pass1, auto pass2) { return comparePassQuality(pass1, pass2); });
-    if (passes_to_optimize.empty())
+    for (ZoneEnum zone_id : pitch_division_->getAllZoneIds())
     {
-        throw std::runtime_error(
-            "passes_to_optimize is empty in PassGenerator, this should never happen");
+        // The objective function we minimize in gradient descent to improve each pass
+        // that we're optimizing
+        const auto objective_function =
+            [this, &world,
+             zone_id](const std::array<double, NUM_PARAMS_TO_OPTIMIZE>& pass_array) {
+                return ratePass(world, Pass::fromPassArray(pass_array),
+                                pitch_division_->getZone(zone_id), passing_config_);
+            };
+
+        auto pass_array = optimizer_.maximize(
```

Comment:
are we using cost functions? If so, why are we maximizing?

Reply:
Yeah this is just a quirk of the pass generator, its been backwards forever
created a ticket to track the work here to fix it: https://github.com/UBC-Thunderbots/Software/issues/2001